stacker/util.py

```diff
@@ -448,3 +452,168 @@ def ensure_s3_bucket(s3_client, bucket_name, bucket_region):
+            ref = config['tag']
+            dir_name = self.sanitize_git_path(uri=config['uri'],
+                                              ref=ref)
+        else:
+            ref = self.git_ls_remote(
+                config['uri'],
+                self.determine_git_ls_remote_ref(config)
+            )
+            dir_name = self.sanitize_git_path(uri=config['uri'],
+                                              ref=ref)
+
+        cached_dir_path = os.path.join(self.package_cache_dir, dir_name)
+
+        # Clone the repo if it doesn't already exist
+        if not os.path.isdir(cached_dir_path):
+            tmp_dir = tempfile.mkdtemp(prefix='stacker')
+            try:
+                tmp_repo_path = os.path.join(tmp_dir, dir_name)
+                with Repo.clone_from(config['uri'], tmp_repo_path) as repo:
+                    repo.head.reference = ref
+                    repo.head.reset(index=True, working_tree=True)
+                shutil.move(tmp_repo_path, self.package_cache_dir)
+            finally:
+                shutil.rmtree(tmp_dir)
+
+        # Cloning (if necessary) is complete. Now add the appropriate
+        # directory (or directories) to sys.path
+        if 'paths' in config:
+            for path in config['paths']:
+                path_to_append = os.path.join(self.package_cache_dir,
+                                              dir_name,
+                                              path)
+                logger.debug("Appending \"%s\" to python sys.path",
+                             path_to_append)
+                sys.path.append(os.path.join(self.package_cache_dir,
+                                             dir_name,
+                                             path))
+        else:
+            sys.path.append(cached_dir_path)
+
+    def git_ls_remote(self, uri, ref):
+        """Determines the latest commit id for a given ref.
+
+        Args:
+            uri (string): git URI
+            ref (string): git ref
+
+        Returns: string (commit id)
+        """
+        logger.debug("Invoking git to retrieve commit id for repo %s...", uri)
+        lsremote_output = subprocess.check_output(['git',
+                                                   'ls-remote',
+                                                   uri,
+                                                   ref])
+        if "\t" in lsremote_output:
+            commit_id = lsremote_output.split("\t")[0]
+            logger.debug("Matching commit id found: %s", commit_id)
+            return commit_id
+        else:
+            raise ValueError("Ref \"%s\" not found for repo %d." % (ref, uri))
+
+    def determine_git_ls_remote_ref(self, config):
+        """Takes a dict describing a git repo and determines the ref to be used
+           with the "git ls-remote" command
+
+        Args:
+            config (dict): git config dictionary; 'branch' key is optional
+
+        Returns: string (a branch reference or "HEAD")
```

Comment:
Can you split this up, and make it more like other docs?  This would be:

```
Returns:
    str: A branch reference or "HEAD"
```

This should happen in a few places in this PR.